benchmark/test/test_driver.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2015 ClusterHQ Inc.  See LICENSE file for details.
+"""
+Driver tests for the control service benchmarks.
+"""
+
+from itertools import count, repeat
+
+from zope.interface import implementer
+
+from twisted.internet.defer import Deferred, succeed, fail
+from twisted.trial.unittest import TestCase
+
+from benchmark._driver import benchmark
+from benchmark._interfaces import IProbe, IOperation
+
+
+class FakeMetric:
+
+    def __init__(self, measurements):
+        """
+        :param measurements: An iterator providing measurement to be
+            returned on each call.
+        """
+        self.measurements = measurements
+
+    def measure(self, f, *a, **kw):
+        def finished(ignored):
+            return next(self.measurements)
+
+        d = f(*a, **kw)
+        d.addCallback(finished)
+        return d
+
+
+@implementer(IProbe)
+class FakeProbe:
+    """
+    A probe performs a single operation, which can be timed.
+    """
+    def __init__(self, good):
+        self.good = good
+
+    def run(self):
+        if self.good:
+            return succeed(None)
+        else:
+            return fail(RuntimeError('ProbeFailed'))
+
+    def cleanup(self):
+        return succeed(None)
+
+
+@implementer(IOperation)
+class FakeOperation:
+
+    def __init__(self, succeeds):
+        """
+        :param succeeds: An iterator providing a boolean indicating
+            whether the probe will succeed.
+        """
+        self.succeeds = succeeds
+
+    def get_probe(self):
+        return FakeProbe(next(self.succeeds))
+
+
+class FakeScenario:
+
+    def __init__(self):
+        self.scenario_maintained = Deferred()
+
+    def start(self):
+        return succeed(None)
+
+    def maintained(self):
+        return self.scenario_maintained
+
+    def stop(self):
+        return succeed(None)
+
+
+class FakeCollapsingScenario:
+
+    def __init__(self):
+        self.scenario_maintained = fail(RuntimeError('collapse'))
+
+    def start(self):
+        return succeed(None)
+
+    def maintained(self):
+        return self.scenario_maintained
+
+    def stop(self):
+        return succeed(None)
+
+
+class SampleTest(TestCase):
+
+    def test_good_probes(self):
+        """
+        Sampling returns results when probes succeed.
+        """
+        samples_ready = benchmark(
+            FakeMetric(count(5)),
+            FakeOperation(repeat(True)),
+            FakeScenario(),
+            3)
+
+        def check(samples):
+            self.assertEqual(
+                samples, [{'success': True, 'value': x} for x in [5, 6, 7]])
+        samples_ready.addCallback(check)
```

Comment:
I tried converting to `SynchronousTestCase`, but the `cooperate` function in `benchmark` uses the global reactor: http://twistedmatrix.com/trac/browser/tags/releases/twisted-15.4.0/twisted/internet/task.py#L337